src/status/condition/mod.rs

```diff
@@ -0,0 +1,472 @@
+
+    old_resource_conditions
+        .merge(new_resource_conditions, update_timestamps)
+        .into()
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ClusterCondition {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Last time the condition transitioned from one status to another.
+    pub last_transition_time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// The last time this condition was updated.
+    pub last_update_time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// A human readable message indicating details about the transition.
+    pub message: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// The reason for the condition's last transition.
+    pub reason: Option<String>,
+    /// Status of the condition, one of True, False, Unknown.
+    pub status: ClusterConditionStatus,
+    /// Type of deployment condition.
+    #[serde(rename = "type")]
+    pub type_: ClusterConditionType,
+}
+
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    Deserialize,
+    EnumCount,
+    Eq,
+    Hash,
+    JsonSchema,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Serialize,
+)]
+#[serde(rename_all = "PascalCase")]
+pub enum ClusterConditionType {
+    #[default]
+    /// Available indicates that the binary maintained by the operator (eg: zookeeper for the
+    /// zookeeper-operator), is functional and available in the cluster.
+    Available,
+    /// Degraded indicates that the operand is not functioning completely. An example of a degraded
+    /// state would be if there should be 5 copies of the operand running but only 4 are running.
+    /// It may still be available, but it is degraded.
+    Degraded,
+    /// Progressing indicates that the operator is actively making changes to the binary maintained
+    /// by the operator (eg: zookeeper for the zookeeper-operator).
+    Progressing,
+    /// Paused indicates that the operator is not reconciling the cluster. This may be used for
+    /// debugging or operator updating.
+    ReconciliationPaused,
+    /// Stopped indicates that all the cluster replicas are scaled down to 0. All resources (e.g.
+    /// ConfigMaps, Services etc.) are kept.
+    Stopped,
+}
+
+#[derive(
+    Clone, Debug, Default, Deserialize, Eq, JsonSchema, Ord, PartialEq, PartialOrd, Serialize,
+)]
+#[serde(rename_all = "PascalCase")]
+pub enum ClusterConditionStatus {
+    #[default]
+    /// True means a resource is in the condition.
+    True,
+    /// False means a resource is not in the condition.
+    False,
+    /// Unknown means kubernetes cannot decide if a resource is in the condition or not.
+    Unknown,
+}
+
+#[derive(Clone, Default)]
+/// Helper struct to order and merge `ClusterCondition` objects.
+pub struct ClusterConditionSet {
+    conditions: Vec<Option<ClusterCondition>>,
```

Comment:
Its a fixed sized vector (size is the amount of fields in `ClusterConditionType`). Must have a "default" value if not set.

Reply:
> couldn't we just not add the element if it is None and not have the Option here?

This would make the merging code much more more complex. Also it would require more code to implement the `set` contract.

Reply:
ok